c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -482,6 +482,32 @@ class CachingAllocatorConfig {
   }
 };
 
+// records allocate/free events
+class MemoryEventTracker {
+ private:
+  std::mutex mutex;
+
+ public:
+  std::vector<std::vector<AllocFreeEvent>> alloc_free_events;
+  void append_alloc_free_event(
+      intptr_t ptr,
+      size_t size,
+      bool type,
+      int device) {
+    std::lock_guard<std::mutex> lock(mutex);
+    alloc_free_events[device].push_back(AllocFreeEvent{
+        ptr, // ptr
+        size, // size: of allocation in bytes
+        type, // type: 1 = allocation; 0 = free
+        device // allocation device
+    });
+  }
+  std::vector<std::vector<AllocFreeEvent>> get_alloc_free_events() const {
+    return alloc_free_events;
```

Comment:
I saw you removed mutex here. You can keep it by making it as mutable.

Reply:
Done. Thank you.